File: packages/core/src/deploy/deploy-orchestrator.ts
// ============================================================
// DeployOrchestrator — the core deploy orchestrator
//
// This is the brain of GitPilot.
// It does not depend on GitHub, VS Code, or IntelliJ.
// It only depends on interfaces. Every platform plugs in by
// implementing those interfaces.
//
// Flow: check changes → filter → Build (optional) → check remote →
//       Stage → Commit → Push → Release (optional)
//
// Principle: on build failure or push conflict, always abort and report
// ============================================================

import type { IDeployOrchestrator } from '../interfaces/IDeployOrchestrator';
import type { IRepositoryProvider } from '../interfaces/IRepositoryProvider';
import type { IGitProvider } from '../interfaces/IGitProvider';
import type { IBuildProvider } from '../interfaces/IBuildProvider';
import type { IReleaseProvider } from '../interfaces/IReleaseProvider';
import type { INotificationProvider } from '../interfaces/INotificationProvider';
import type { DeployConfig, DeployResult, DeployStep, DeployStatus } from '../models/deploy';
import { Logger } from '../utils/logger';

export class DeployOrchestrator implements IDeployOrchestrator {
  private logger = new Logger('Orchestrator');
  private currentStatus: DeployStatus = 'idle';
  private onStatusChange?: (status: DeployStatus, step: DeployStep) => void;

  constructor(
    private readonly config: DeployConfig,
    private readonly repoProvider: IRepositoryProvider,
    private readonly gitProvider: IGitProvider,
    private readonly buildProvider: IBuildProvider,
    private readonly releaseProvider: IReleaseProvider,
    private readonly notifier: INotificationProvider,
    private readonly smartFilter: { filter: (files: string[]) => { allowed: string[]; blocked: Array<{ file: string; reason: string }> } },
  ) {
    this.config = {
      commitMessageTemplate: 'deploy: auto-deploy by GitPilot',
      buildBeforeDeploy: false,
      blockOnBuildFailure: true,
      ...config,
    };
  }

  onStepChange(callback: (status: DeployStatus, step: DeployStep) => void): void {
    this.onStatusChange = callback;
  }

  getStatus(): DeployStatus {
    return this.currentStatus;
  }

  async hasPendingChanges(): Promise<boolean> {
    return this.gitProvider.hasChanges();
  }

  // ================================================================
  //  Core deploy flow
  // ================================================================

  async deploy(commitMessage?: string): Promise<DeployResult> {
    const startTime = Date.now();
    const steps: DeployStep[] = [];
    const message = commitMessage ?? this.config.commitMessageTemplate ?? 'deploy: auto-deploy by GitPilot';

    try {
      // ── Step 1: Check for changes ──
      const stepCheck = this.createStep('Checking file changes');
      this.emit('staging', stepCheck);

      const hasChanges = await this.gitProvider.hasChanges();
      const gitStatus = await this.gitProvider.getStatus();
      const hasUnpushedCommits = (gitStatus.ahead ?? 0) > 0;

      // ⭐ Neither file changes nor unpushed commits → skip
      if (!hasChanges && !hasUnpushedCommits) {
        stepCheck.status = 'skipped';
        stepCheck.details = 'No changes to deploy';
        this.finishStep(stepCheck);
        steps.push(stepCheck);
        this.currentStatus = 'no-changes';
        return { success: true, status: 'no-changes', steps, totalDurationMs: Date.now() - startTime };
      }

      // ⭐ Unpushed commits but no file changes → skip staging/commit, push directly
      if (!hasChanges && hasUnpushedCommits) {
        stepCheck.status = 'success';
        stepCheck.details = `No new changes, but ${gitStatus.ahead} commit(s) not pushed`;
        this.finishStep(stepCheck);
        steps.push(stepCheck);

        // Jump straight to push
        const stepPush = this.createStep('Pushing to remote');
        this.emit('pushing', stepPush);
        const pushResult = await this.gitProvider.push('origin', this.config.branch);
        if (!pushResult.success) {
          stepPush.status = 'failed';
          stepPush.error = pushResult.error;
          this.finishStep(stepPush);
          steps.push(stepPush);
          this.currentStatus = 'failed';
          return {
            success: false, status: 'failed', steps,
            error: `Push failed: ${pushResult.error}`,
            totalDurationMs: Date.now() - startTime,
          };
        }
        stepPush.status = 'success';
        stepPush.details = 'Pushed';
        this.finishStep(stepPush);
        steps.push(stepPush);
        this.currentStatus = 'success';
        return { success: true, status: 'success', steps, totalDurationMs: Date.now() - startTime };
      }

      stepCheck.status = 'success';
      this.finishStep(stepCheck);
      steps.push(stepCheck);

      // ── Step 2: Smart filtering ──
      const stepFilter = this.createStep('Smart file filtering');
      this.emit('staging', stepFilter);

      const allChanged = [
        ...gitStatus.modified,
        ...gitStatus.added,
        ...gitStatus.deleted,
        ...gitStatus.untracked,
      ];
      const filterResult = this.smartFilter.filter(allChanged);

      if (filterResult.blocked.length > 0) {
        this.logger.warn(`Filtered out ${filterResult.blocked.length} file(s)`);
      }
      stepFilter.status = 'success';
      stepFilter.details = `${filterResult.allowed.length} file(s) to deploy`;
      if (filterResult.blocked.length > 0) {
        stepFilter.details += `, filtered out ${filterResult.blocked.length}`;
      }
      this.finishStep(stepFilter);
      steps.push(stepFilter);

      // ── Step 3: Build (optional) ──
      if (this.config.buildBeforeDeploy && this.config.buildCommand) {
        const stepBuild = this.createStep('Building project');
        this.emit('building', stepBuild);

        const buildResult = await this.buildProvider.runBuild({
          command: this.config.buildCommand,
          timeout: 5 * 60 * 1000,
        });

        if (!buildResult.success) {
          stepBuild.status = 'failed';
          stepBuild.error = buildResult.error ?? 'Build failed';
          this.finishStep(stepBuild);
          stepBuild.durationMs = buildResult.durationMs;
          steps.push(stepBuild);

          if (this.config.blockOnBuildFailure) {
            this.currentStatus = 'failed';
            return {
              success: false, status: 'failed', steps,
              error: `Build failed: ${buildResult.error}`,
              totalDurationMs: Date.now() - startTime,
            };
          }
        }

        stepBuild.status = 'success';
        stepBuild.details = buildResult.output?.slice(-200);
        this.finishStep(stepBuild);
        stepBuild.durationMs = buildResult.durationMs;
        steps.push(stepBuild);
      }

      // ── Step 4: Check remote status ──
      const stepRemote = this.createStep('Checking remote status');
      this.emit('staging', stepRemote);

      const branch = this.config.branch;
      const aheadCount = await this.repoProvider.checkRemoteAhead(
        this.config.repo.owner,
        this.config.repo.name,
        branch,
      );

      if (aheadCount > 0) {
        stepRemote.status = 'failed';
        stepRemote.error = `Remote is ahead by ${aheadCount} commit(s); please run Sync first`;
        this.finishStep(stepRemote);
        steps.push(stepRemote);
        this.currentStatus = 'failed';
        return {
          success: false, status: 'failed', steps,
          error: 'The remote has new commits (non-fast-forward); push aborted. Please run Sync first.',
          totalDurationMs: Date.now() - startTime,
        };
      }
      stepRemote.status = 'success';
      stepRemote.details = 'Remote status OK';
      this.finishStep(stepRemote);
      steps.push(stepRemote);

      // ── Step 5: Stage ──
      const stepStage = this.createStep('Staging files');
      this.emit('staging', stepStage);
      // ⭐ git add -A everything → exclude blocked files (secrets/build artifacts, etc.)
      await this.gitProvider.stageFiles(filterResult.blocked.map((b) => b.file));
      // ⭐ Verify the staging result
      const stagedFiles = await this.gitProvider.getStagedFiles();
      stepStage.status = 'success';
      stepStage.details = stagedFiles.length > 0
        ? `Staged ${stagedFiles.length} file(s)`
        : `${filterResult.allowed.length} file(s) to commit`;
      if (filterResult.blocked.length > 0) {
        stepStage.details += `, excluded ${filterResult.blocked.length}`;
      }
      this.finishStep(stepStage);
      steps.push(stepStage);

      // If nothing was staged, skip the commit
      if (stagedFiles.length === 0 && filterResult.allowed.length === 0) {
        this.currentStatus = 'no-changes';
        return { success: true, status: 'no-changes', steps, totalDurationMs: Date.now() - startTime };
      }

      // ── Step 6: Commit ──
      const stepCommit = this.createStep('Committing changes');
      this.emit('committing', stepCommit);
      const commitResult = await this.gitProvider.commit(message);

      if (!commitResult.success) {
        stepCommit.status = 'failed';
        stepCommit.error = commitResult.error;
        this.finishStep(stepCommit);
        steps.push(stepCommit);
        this.currentStatus = 'failed';
        return {
          success: false, status: 'failed', steps,
          error: `Commit failed: ${commitResult.error}`,
          totalDurationMs: Date.now() - startTime,
        };
      }
      stepCommit.status = 'success';
      stepCommit.details = `${commitResult.hash?.substring(0, 7)} - ${message}`;
      this.finishStep(stepCommit);
      steps.push(stepCommit);

      // ── Step 7: Push ──
      const stepPush = this.createStep('Pushing to remote');
      this.emit('pushing', stepPush);
      const pushResult = await this.gitProvider.push('origin', branch);

      if (!pushResult.success) {
        stepPush.status = 'failed';
        stepPush.error = pushResult.error;
        this.finishStep(stepPush);
        steps.push(stepPush);
        this.currentStatus = 'failed';
        return {
          success: false, status: 'failed', steps,
          commitHash: commitResult.hash ?? undefined,
          error: pushResult.nonFastForward
            ? 'Push conflict: the remote has new commits, please run Sync first'
            : `Push failed: ${pushResult.error}`,
          totalDurationMs: Date.now() - startTime,
        };
      }
      stepPush.status = 'success';
      stepPush.details = pushResult.pushed ? 'Pushed' : 'Remote is already up to date';
      this.finishStep(stepPush);
      steps.push(stepPush);

      // ── Step 8: Release (optional) ──
      let releaseUrl: string | undefined;
      if (this.config.release?.enabled) {
        const stepRelease = this.createStep('Creating Release');
        this.emit('releasing', stepRelease);

        try {
          const tagName = this.generateTagName();
          const release = await this.releaseProvider.createRelease(
            this.config.repo.owner,
            this.config.repo.name,
            {
              tagName,
              name: tagName,
              prerelease: this.config.release.prerelease ?? false,
              draft: this.config.release.draft ?? false,
              targetCommitish: commitResult.hash ?? undefined,
            },
          );
          releaseUrl = release.htmlUrl;
          stepRelease.status = 'success';
          stepRelease.details = tagName;
          this.finishStep(stepRelease);
          steps.push(stepRelease);
        } catch (error: any) {
          stepRelease.status = 'failed';
          stepRelease.error = error.message;
          this.finishStep(stepRelease);
          steps.push(stepRelease);
          this.logger.warn(`Release creation failed (does not block the deploy): ${error.message}`);
        }
      }

      // ── Done ──
      this.currentStatus = 'success';
      const totalMs = Date.now() - startTime;
      this.logger.info(`Deploy finished in ${(totalMs / 1000).toFixed(1)}s`);

      return {
        success: true,
        status: 'success',
        steps,
        commitHash: commitResult.hash ?? undefined,
        releaseUrl,
        totalDurationMs: totalMs,
      };
    } catch (error: any) {
      this.currentStatus = 'failed';
      this.logger.error(`Deploy error: ${error.message}`);
      return {
        success: false,
        status: 'failed',
        steps,
        error: error.message,
        totalDurationMs: Date.now() - startTime,
      };
    }
  }

  // ================================================================
  //  Sync flow
  // ================================================================

  async sync(): Promise<DeployResult> {
    const startTime = Date.now();
    const steps: DeployStep[] = [];

    try {
      const stepPull = this.createStep('Pulling remote updates');
      this.emit('idle', stepPull);

      const pullResult = await this.gitProvider.pull();
      stepPull.status = pullResult.success ? 'success' : 'failed';
      stepPull.error = pullResult.error;
      this.finishStep(stepPull);
      steps.push(stepPull);

      const deployResult = await this.deploy();
      steps.push(...deployResult.steps);

      return { ...deployResult, steps, totalDurationMs: Date.now() - startTime };
    } catch (error: any) {
      return {
        success: false, status: 'failed', steps,
        error: error.message,
        totalDurationMs: Date.now() - startTime,
      };
    }
  }

  // ================================================================
  //  Internal helpers
  // ================================================================

  private createStep(name: string): DeployStep {
    return { name, status: 'running', startTime: new Date() };
  }

  private finishStep(step: DeployStep): void {
    step.endTime = new Date();
    if (step.startTime) {
      step.durationMs = step.endTime.getTime() - step.startTime.getTime();
    }
  }

  private emit(status: DeployStatus, step: DeployStep): void {
    this.currentStatus = status;
    this.onStatusChange?.(status, step);
  }

  private generateTagName(): string {
    const now = new Date();
    const y = now.getFullYear();
    const m = String(now.getMonth() + 1).padStart(2, '0');
    const d = String(now.getDate()).padStart(2, '0');
    const h = String(now.getHours()).padStart(2, '0');
    const min = String(now.getMinutes()).padStart(2, '0');
    return `v${y}${m}${d}-${h}${min}`;
  }
}

File: packages/jetbrains-plugin/src/main/kotlin/com/gitpilot/core/DeployOrchestrator.kt
// ============================================================
// GitPilot JetBrains — DeployOrchestrator (Kotlin) v1.2.9 sync
// Fully aligned with every feature of the VS Code deploy-orchestrator.ts
// Including: unpushed-commit detection, branch verification, file filtering, Release
// ============================================================

package com.gitpilot.core

import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.Project

// ── Data classes ──

enum class DeployStatus { IDLE, BUILDING, STAGING, COMMITTING, PUSHING, RELEASING, SUCCESS, FAILED, NO_CHANGES }

data class DeployStep(
    val name: String,
    var status: String = "pending",
    var error: String? = null,
    var details: String? = null,
    var startTime: Long = System.currentTimeMillis(),
    var endTime: Long? = null,
    var durationMs: Long? = null
) {
    fun finish() { endTime = System.currentTimeMillis(); durationMs = endTime!! - startTime }
}

data class DeployResult(
    val success: Boolean,
    val status: DeployStatus,
    val steps: List<DeployStep>,
    val commitHash: String? = null,
    val releaseUrl: String? = null,
    val error: String? = null,
    val durationMs: Long
)

data class FilterResult(
    val allowed: List<String>,
    val blocked: List<Pair<String, String>>
)

// ── Interfaces (full version, aligned with the TS interfaces/) ──

interface IRepoProvider {
    val platform: String
    fun listRepos(): List<GitHubRepo>
    fun createRepo(name: String, isPrivate: Boolean): GitHubRepo?
    fun deleteRepo(owner: String, repo: String): Boolean
}

interface IGitProvider {
    fun hasChanges(): Boolean
    fun getStatus(): GitStatus?
    fun stageFiles(excludeFiles: List<String> = emptyList()): Boolean
    fun getStagedFiles(): List<String>
    fun commit(message: String): CommitResult
    fun push(remote: String = "origin"): PushResult
    fun pull(): Boolean
    fun getCurrentBranch(): String?
    fun getRemoteUrl(remote: String = "origin"): String?
    fun hasRemote(remote: String = "origin"): Boolean
    fun addRemote(name: String, url: String): Boolean
}

interface IBuildProvider {
    fun run(command: String, indicator: ProgressIndicator): BuildResult
}

interface IReleaseProvider {
    fun createRelease(owner: String, repo: String, tag: String, name: String, commitHash: String?, prerelease: Boolean, draft: Boolean): ReleaseResult?
}

// ── Data classes ──

data class GitHubRepo(val id: Long, val name: String, val fullName: String, val owner: String, val isPrivate: Boolean, val htmlUrl: String, val cloneUrl: String, val defaultBranch: String)
data class GitStatus(val isClean: Boolean, val modified: List<String>, val added: List<String> = emptyList(), val deleted: List<String> = emptyList(), val untracked: List<String> = emptyList(), val currentBranch: String, val ahead: Int = 0, val behind: Int = 0)
data class CommitResult(val success: Boolean, val hash: String?, val error: String?)
data class PushResult(val success: Boolean, val pushed: Boolean, val error: String?, val nonFastForward: Boolean = false)
data class BuildResult(val success: Boolean, val output: String?, val error: String?, val durationMs: Long)
data class ReleaseResult(val htmlUrl: String, val tag: String)

// ── DeployOrchestrator v1.2.9 ──

@Service(Service.Level.PROJECT)
class DeployOrchestrator(private val project: Project) {

    private val logger = Logger.getInstance(DeployOrchestrator::class.java)
    var currentStatus: DeployStatus = DeployStatus.IDLE

    fun deploy(
        indicator: ProgressIndicator,
        repoProvider: IRepoProvider,
        gitProvider: IGitProvider,
        buildProvider: IBuildProvider?,
        releaseProvider: IReleaseProvider?,
        buildCommand: String?,
        buildBeforeDeploy: Boolean,
        blockOnBuildFailure: Boolean,
        smartFilter: SmartFilter?,
        commitMessage: String = "deploy: auto-deploy by GitPilot"
    ): DeployResult {
        val startTime = System.currentTimeMillis()
        val steps = mutableListOf<DeployStep>()

        try {
            // ═══ Step 1: Check for changes ═══
            indicator.text = "Checking file changes..."
            indicator.fraction = 0.1
            val stepCheck = DeployStep("Checking file changes")
            val hasChanges = gitProvider.hasChanges()
            val gitStatus = gitProvider.getStatus()
            val hasUnpushedCommits = (gitStatus?.ahead ?: 0) > 0

            // Neither changes nor unpushed commits → skip
            if (!hasChanges && !hasUnpushedCommits) {
                stepCheck.status = "skipped"; stepCheck.details = "No changes to deploy"; stepCheck.finish(); steps.add(stepCheck)
                currentStatus = DeployStatus.NO_CHANGES
                return DeployResult(true, DeployStatus.NO_CHANGES, steps, durationMs = System.currentTimeMillis() - startTime)
            }

            // Unpushed commits but no new changes → push directly
            if (!hasChanges && hasUnpushedCommits) {
                stepCheck.status = "success"; stepCheck.details = "No new changes, ${gitStatus!!.ahead} commit(s) not pushed"; stepCheck.finish(); steps.add(stepCheck)
                return doPushOnly(indicator, gitProvider, steps, startTime)
            }

            stepCheck.status = "success"; stepCheck.finish(); steps.add(stepCheck)

            // ═══ Step 2: Smart filtering ═══
            indicator.text = "Smart file filtering..."
            indicator.fraction = 0.2
            val stepFilter = DeployStep("Smart file filtering")
            val allChanged = mutableListOf<String>()
            gitStatus?.let {
                allChanged.addAll(it.modified); allChanged.addAll(it.added)
                allChanged.addAll(it.deleted); allChanged.addAll(it.untracked)
            }
            val filterResult = smartFilter?.filter(allChanged) ?: FilterResult(allChanged, emptyList())
            stepFilter.status = "success"
            stepFilter.details = "${filterResult.allowed.size} file(s) to deploy"
            if (filterResult.blocked.isNotEmpty()) stepFilter.details += ", excluded ${filterResult.blocked.size}"
            stepFilter.finish(); steps.add(stepFilter)

            // ═══ Step 3: Build (optional) ═══
            if (buildBeforeDeploy && buildCommand != null && buildProvider != null) {
                indicator.text = "Building project..."
                indicator.fraction = 0.35
                val stepBuild = DeployStep("Building project")
                val result = buildProvider.run(buildCommand, indicator)
                stepBuild.status = if (result.success) "success" else "failed"
                stepBuild.error = result.error; stepBuild.details = result.output?.takeLast(200)
                stepBuild.finish(); steps.add(stepBuild)
                if (!result.success && blockOnBuildFailure) {
                    currentStatus = DeployStatus.FAILED
                    return DeployResult(false, DeployStatus.FAILED, steps, error = "Build failed: ${result.error}", durationMs = System.currentTimeMillis() - startTime)
                }
            }

            // ═══ Step 4: Stage ═══
            indicator.text = "Staging files..."
            indicator.fraction = 0.5
            val stepStage = DeployStep("Staging files")
            val excludedPaths = filterResult.blocked.map { it.first }
            val staged = gitProvider.stageFiles(excludedPaths)
            val stagedFiles = gitProvider.getStagedFiles()
            stepStage.status = if (staged) "success" else "failed"
            stepStage.details = if (stagedFiles.isNotEmpty()) "Staged ${stagedFiles.size} file(s)" else "${filterResult.allowed.size} file(s) to commit"
            if (!staged) stepStage.error = "Staging failed"
            stepStage.finish(); steps.add(stepStage)
            if (!staged) return DeployResult(false, DeployStatus.FAILED, steps, error = "Staging failed", durationMs = System.currentTimeMillis() - startTime)
            if (stagedFiles.isEmpty() && filterResult.allowed.isEmpty()) {
                currentStatus = DeployStatus.NO_CHANGES
                return DeployResult(true, DeployStatus.NO_CHANGES, steps, durationMs = System.currentTimeMillis() - startTime)
            }

            // ═══ Step 5: Commit ═══
            indicator.text = "Committing changes..."
            indicator.fraction = 0.65
            val stepCommit = DeployStep("Committing changes")
            val commitResult = gitProvider.commit(commitMessage)
            stepCommit.status = if (commitResult.success) "success" else "failed"
            stepCommit.error = commitResult.error; stepCommit.details = commitResult.hash?.take(7)
            stepCommit.finish(); steps.add(stepCommit)
            if (!commitResult.success) return DeployResult(false, DeployStatus.FAILED, steps, error = "Commit failed", durationMs = System.currentTimeMillis() - startTime)

            // ═══ Step 6: Push ═══
            return doPush(indicator, gitProvider, releaseProvider, commitResult.hash, steps, startTime)

        } catch (e: Exception) {
            logger.error("Deploy failed", e)
            currentStatus = DeployStatus.FAILED
            return DeployResult(false, DeployStatus.FAILED, steps, error = e.message, durationMs = System.currentTimeMillis() - startTime)
        }
    }

    /** Push only (when there are no new changes) */
    private fun doPushOnly(indicator: ProgressIndicator, gitProvider: IGitProvider, steps: MutableList<DeployStep>, startTime: Long): DeployResult {
        indicator.text = "Pushing to GitHub..."; indicator.fraction = 0.7
        val stepPush = DeployStep("Pushing to GitHub")
        val pushResult = gitProvider.push()
        stepPush.status = if (pushResult.success) "success" else "failed"
        stepPush.error = pushResult.error; stepPush.details = if (pushResult.pushed) "Pushed" else "Remote is already up to date"
        stepPush.finish(); steps.add(stepPush)
        if (!pushResult.success) {
            val msg = if (pushResult.nonFastForward) "Push conflict: the remote has new commits, please run Sync first" else "Push failed: ${pushResult.error}"
            return DeployResult(false, DeployStatus.FAILED, steps, error = msg, durationMs = System.currentTimeMillis() - startTime)
        }
        currentStatus = DeployStatus.SUCCESS
        return DeployResult(true, DeployStatus.SUCCESS, steps, durationMs = System.currentTimeMillis() - startTime)
    }

    /** Full push + Release */
    private fun doPush(indicator: ProgressIndicator, gitProvider: IGitProvider, releaseProvider: IReleaseProvider?, commitHash: String?, steps: MutableList<DeployStep>, startTime: Long): DeployResult {
        indicator.text = "Pushing to GitHub..."; indicator.fraction = 0.8
        val stepPush = DeployStep("Pushing to GitHub")
        val pushResult = gitProvider.push()
        stepPush.status = if (pushResult.success) "success" else "failed"
        stepPush.error = pushResult.error; stepPush.details = if (pushResult.pushed) "Pushed" else "Remote is already up to date"
        stepPush.finish(); steps.add(stepPush)
        if (!pushResult.success) {
            val msg = if (pushResult.nonFastForward) "Push conflict: the remote has new commits, please run Sync first" else "Push failed: ${pushResult.error}"
            return DeployResult(false, DeployStatus.FAILED, steps, commitHash = commitHash, error = msg, durationMs = System.currentTimeMillis() - startTime)
        }

        // ═══ Step 7: Release (optional) ═══
        var releaseUrl: String? = null
        if (releaseProvider != null) {
            indicator.text = "Creating Release..."; indicator.fraction = 0.95
            val stepRelease = DeployStep("Creating Release")
            try {
                val tag = "v${System.currentTimeMillis() / 1000}"
                val release = releaseProvider.createRelease("unknown", "unknown", tag, tag, commitHash, false, false)
                if (release != null) {
                    releaseUrl = release.htmlUrl; stepRelease.status = "success"; stepRelease.details = tag
                } else { stepRelease.status = "skipped"; stepRelease.details = "Release not enabled" }
            } catch (e: Exception) { stepRelease.status = "failed"; stepRelease.error = e.message }
            stepRelease.finish(); steps.add(stepRelease)
        }

        currentStatus = DeployStatus.SUCCESS
        return DeployResult(true, DeployStatus.SUCCESS, steps, commitHash = commitHash, releaseUrl = releaseUrl, durationMs = System.currentTimeMillis() - startTime)
    }

    fun sync(
        indicator: ProgressIndicator,
        repoProvider: IRepoProvider,
        gitProvider: IGitProvider,
        buildProvider: IBuildProvider?,
        releaseProvider: IReleaseProvider?,
        buildCommand: String?,
        buildBeforeDeploy: Boolean,
        blockOnBuildFailure: Boolean,
        smartFilter: SmartFilter?
    ): DeployResult {
        indicator.text = "Pulling remote updates..."
        gitProvider.pull()
        return deploy(indicator, repoProvider, gitProvider, buildProvider, releaseProvider, buildCommand, buildBeforeDeploy, blockOnBuildFailure, smartFilter)
    }
}
